src/coreclr/jit/assertionprop.cpp

```diff
@@ -4863,9 +4863,28 @@ bool Compiler::optNonNullAssertionProp_Ind(ASSERT_VALARG_TP assertions, GenTree*
 {
     assert(indir->OperIsIndir());
 
-    if (!(indir->gtFlags & GTF_EXCEPT))
+    bool updated = false;
+    if (indir->OperIs(GT_STOREIND))
     {
-        return false;
+        // We like CSE to happen for handles, as the codegen for loading a 64-bit constant can be pretty heavy
+        // and this is particularly true on platforms with a fixed-width instruction encoding. However, this
+        // pessimizes stores as we can no longer optimize around some object handles that would allow us to
+        // bypass the write barrier.
+        //
+        // In order to handle that, we'll propagate the IND_TGT_NOT_HEAP flag onto the store if the handle is
+        // directly or if the underlying value number is an applicable object handle.
+
+        const GenTree* data = indir->AsIndir()->Data();
+        if (data->IsIconHandle(GTF_ICON_OBJ_HDL) || vnStore->IsVNObjHandle(data->gtVNPair.GetConservative()))
+        {
+            indir->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+            updated = true;
+        }
+    }
+
+    if ((indir->gtFlags & GTF_EXCEPT) == 0)
+    {
+        return updated;
```

Comment:
Can you do `return updated;` for the bottom `return false;` in this method. Otherwise you modify a tree and `return false;`

Also, I think we need to update comment in `GTF_IND_TGT_NOT_HEAP` because techincally frozen (or nongc) objects **are** in heap, just not in the one managed by GC. Something like `GT_IND -- the target is on the heap managed by GC`

Reply:
I extracted it to its own `optNonHeapAssertionProp_Ind` method to avoid potential confusion or user bugs related to `optNonNullAssertionProp_Ind` and the difference in expectation between the two.

Most notably, that `NonNull` is tracking whether it believes `Addr()` can be null while `NonHeap` is tracking whether the `Data()` can be on a GC-tracked heap.

The comment around `GTF_IND_TGT_NOT_HEAP` was then also updated to indicate that it means the target is not GC-tracked